ic-task-scheduler/src/scheduler.rs

```diff
@@ -1,86 +1,170 @@
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
 
-use ic_stable_structures::UnboundedMapStructure;
+use ic_kit::RejectionCode;
+use ic_stable_structures::{BTreeMapStructure as _, HeapBTreeMap, IterableUnboundedMapStructure};
 use parking_lot::Mutex;
 
 use crate::task::{ScheduledTask, Task};
 use crate::time::time_secs;
-use crate::SchedulerError;
+use crate::{Result, SchedulerError};
+
+/// Internal type used to store tasks to be processed and processing tasks
+type TaskQueue = Arc<Mutex<HeapBTreeMap<u32, (), ()>>>;
+
+/// The state of a task execution.
+/// This is reported when the `SaveStateQueryCallback` is called.
+#[derive(Debug)]
+pub enum TaskExecutionState {
+    /// Reported when tasks to be executed are scheduled
+    Scheduled,
+    /// Reported when a task fails.
+    Failed(u32, SchedulerError),
+    /// Reported when a task starts executing.
+    Executing(u32),
+    /// Reported when a task completes successfully.
+    Completed(u32),
+    /// Reported when a task panics.
+    Panicked(u32),
+}
 
-type SchedulerErrorCallback<T> = Box<dyn 'static + Fn(ScheduledTask<T>, SchedulerError) + Send>;
+type SaveStateQueryCallback = dyn Fn(
+        TaskExecutionState,
+    ) -> Pin<Box<dyn Future<Output = std::result::Result<(), (RejectionCode, String)>>>>
+    + Send
+    + Sync;
 
 /// A scheduler is responsible for executing tasks.
-pub struct Scheduler<T: 'static + Task, P: 'static + UnboundedMapStructure<u32, ScheduledTask<T>>> {
+pub struct Scheduler<T, P>
+where
+    T: 'static + Task,
+    P: 'static + IterableUnboundedMapStructure<u32, ScheduledTask<T>>,
+{
     pending_tasks: Arc<Mutex<P>>,
     phantom: std::marker::PhantomData<T>,
-    failed_task_callback: Arc<Option<SchedulerErrorCallback<T>>>,
+    /// Queue containing tasks to be processed
+    tasks_to_be_processed: TaskQueue,
+    /// Tasks which are currently being processed
+    tasks_running: TaskQueue,
+    /// Callback to be called to save the current canister state to prevent panicking tasks.
+    on_execution_state_changed_callback: Arc<Box<SaveStateQueryCallback>>,
 }
 
-impl<T: 'static + Task, P: 'static + UnboundedMapStructure<u32, ScheduledTask<T>>> Scheduler<T, P> {
+impl<T, P> Scheduler<T, P>
+where
+    T: 'static + Task,
+    P: 'static + IterableUnboundedMapStructure<u32, ScheduledTask<T>>,
+{
     /// Create a new scheduler.
-    pub fn new(pending_tasks: P) -> Self {
-        Self {
+    ///
+    /// A callback must be passed and it'll be called to save and report the current state of the scheduler.
```

Comment:
The description is confusing.
It should say that:
1. this is a callback called every time the state of a task changes
2. By performing an intercanister call in the callback, you can force the state to be persisted even in case of panics. this allows the scheduler to deal with panics